ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java

```diff
@@ -306,22 +314,71 @@ private void recover(MessageStoreRecovery recovery) throws StoreException, IOExc
    * @throws IOException if an end offset could not be set due to I/O error.
    */
   private void setEndOffsets() throws IOException {
-    for (Offset segmentStartOffset : indexes.keySet()) {
-      Offset nextIndexSegmentStartOffset = indexes.higherKey(segmentStartOffset);
+    for (Offset segmentStartOffset : indexesRef.get().keySet()) {
+      Offset nextIndexSegmentStartOffset = indexesRef.get().higherKey(segmentStartOffset);
       if (nextIndexSegmentStartOffset == null || !segmentStartOffset.getName()
           .equals(nextIndexSegmentStartOffset.getName())) {
         // this is the last index segment for the log segment it refers to.
-        IndexSegment indexSegment = indexes.get(segmentStartOffset);
+        IndexSegment indexSegment = indexesRef.get().get(segmentStartOffset);
         LogSegment logSegment = log.getSegment(segmentStartOffset.getName());
         logSegment.setEndOffset(indexSegment.getEndOffset().getOffset());
       }
     }
   }
 
   /**
+   * @return the map of {@link Offset} to {@link IndexSegment} instances.
+   */
+  ConcurrentSkipListMap<Offset, IndexSegment> getIndexSegments() {
+    return indexesRef.get();
+  }
+
+  /**
+   * Atomically adds {@code segmentFilesToAdd} to and removes {@code segmentsToRemove} from the map of {@link Offset} to
+   * {@link IndexSegment} instances.
+   * @param segmentFilesToAdd the backing files of the {@link IndexSegment} instances to add.
+   * @param segmentsToRemove the start {@link Offset} of {@link IndexSegment} instances to remove.
+   * @throws IllegalArgumentException if any {@link IndexSegment} that needs to be added or removed has an offset higher
+   * than that in the journal.
+   * @throws StoreException if an {@link IndexSegment} instance cannot be created for the provided files to add.
+   */
+  void changeIndexSegments(List<File> segmentFilesToAdd, Set<Offset> segmentsToRemove) throws StoreException {
+    TreeMap<Offset, IndexSegment> segmentsToAdd = new TreeMap<>();
+    Offset journalFirstOffset = journal.getFirstOffset();
+    for (File indexSegmentFile : segmentFilesToAdd) {
+      IndexSegment indexSegment = new IndexSegment(indexSegmentFile, true, factory, config, metrics, journal);
+      if (indexSegment.getEndOffset().compareTo(journalFirstOffset) > 0) {
+        throw new IllegalArgumentException("One of the index segments has an end offset " + indexSegment.getEndOffset()
+            + " that is higher than the first offset in the journal " + journalFirstOffset);
+      }
+      segmentsToAdd.put(indexSegment.getStartOffset(), indexSegment);
+    }
+
+    indexesChangeLock.lock();
+    try {
+      ConcurrentSkipListMap<Offset, IndexSegment> oldIndexSegments = indexesRef.get();
+      ConcurrentSkipListMap<Offset, IndexSegment> newIndexSegments = new ConcurrentSkipListMap<>();
+      for (Map.Entry<Offset, IndexSegment> entry : oldIndexSegments.entrySet()) {
+        if (!segmentsToRemove.contains(entry.getKey())) {
```

Comment:
Just to recap what we discussed and concluded in person. 
- The lock is likely avoidable. If the code gets too complicated, we could go with the lock as even though `addToIndex()` is in the critical path, it is called by the store after taking a store lock, so the contention for this lock will not be an issue. If not, we should get rid of the lock as we rarely use locks within the index (if at all) and because we use concurrent data structures and atomic references everywhere and generally follow the lock-free paradigm within the index.
- `ConcurrentSkipListMap` provides weakly consistent iterators which can be used to get a snapshot view of the old map at the point of creation or since creating it. What will need to be ensured is that the happened-before relation between the start of this iteration and `addToIndex()` adding to both old and new references, is maintained.
- We also discussed whether we would really need atomic references. The original reason was to save off the value at the start of methods that need to look at the segments map. This may be possible even without atomic references.